src/bgw/scheduler.c

```diff
@@ -0,0 +1,402 @@
+#include <utils/memutils.h>
+#include <access/xact.h>
+#include <pgstat.h>
+
+#include "extension.h"
+#include "scheduler.h"
+#include "job.h"
+#include "job_stat.h"
+#include "version.h"
+#include "compat.h"
+
+#define MAX_TIMEOUT (5*1000L)
+#define SCHEDULER_APPNAME "TimescaleDB Background Worker Scheduler"
+
+TS_FUNCTION_INFO_V1(ts_bgw_scheduler_main);
+
+
+/* has to be global to shutdowns jobs on exit */
+static List *scheduled_jobs = NIL;
+
+typedef struct ScheduledBgwJob
+{
+	BgwJob	   *job;
+	TimestampTz next_start;
+	TimestampTz timeout_at;
+
+	bool		enabled;
+	BackgroundWorkerHandle *handle;
+} ScheduledBgwJob;
+
+static void scheduled_bgw_job_mark_stopped(ScheduledBgwJob *job);
+
+static inline void
+bgw_scheduler_on_postmaster_death(void)
+{
+	/* don't call exit hooks cause we want to bail out quickly */
+	on_exit_reset();
+	ereport(FATAL,
+			(errcode(ERRCODE_ADMIN_SHUTDOWN),
+			 errmsg("postmaster exited while timescaledb scheduler was working")));
+}
+static long
+get_timeout(TimestampTz by_time)
+{
+	long		timeout_sec = 5;
+	int			timeout_usec = 0;
+
+	if (TIMESTAMP_IS_NOBEGIN(by_time))
+		return 0;
+
+	TimestampDifference(GetCurrentTimestamp(), by_time, &timeout_sec, &timeout_usec);
+
+
+	if (timeout_sec <= 0 && timeout_usec <= 0)
+		return 0;
+
+	return timeout_sec * 1000L + timeout_usec / 1000;
+}
+
+static void
+scheduled_bgw_job_start_worker(ScheduledBgwJob *sjob)
+{
+
+	pid_t		pid;
+	BgwHandleStatus status;
+
+
+	Assert(sjob->enabled);
+
+	StartTransactionCommand();
+	ts_bgw_job_stat_mark_start(sjob->job->fd.id);
+	if (ts_bgw_job_has_timeout(sjob->job))
+		sjob->timeout_at = ts_bgw_job_timeout_at(sjob->job, GetCurrentTransactionStartTimestamp());
+	else
+		sjob->timeout_at = DT_NOEND;
+	CommitTransactionCommand();
+
+	elog(DEBUG1, "scheduling job %d \"%s\"", sjob->job->fd.id, NameStr(sjob->job->fd.job_name));
+
+
+	sjob->handle = ts_bgw_job_start_worker(sjob->job);
+	status = WaitForBackgroundWorkerStartup(sjob->handle, &pid);
+	switch (status)
+	{
+		case BGWH_POSTMASTER_DIED:
+			bgw_scheduler_on_postmaster_death();
+			break;
+		case BGWH_STARTED:
+			/* all good */
+			break;
+		case BGWH_STOPPED:
+			sjob->handle = NULL;
+			StartTransactionCommand();
+			ts_bgw_job_stat_mark_end(sjob->job, false);
+			scheduled_bgw_job_mark_stopped(sjob);
+			CommitTransactionCommand();
+			break;
+		case BGWH_NOT_YET_STARTED:
+			/* should not be possible */
+			elog(ERROR, "unexpected bgworker state %d", status);
+			break;
+	}
+}
+
+static ScheduledBgwJob *
+scheduled_bgw_job_create(BgwJob *job, MemoryContext mctx)
+{
+	BgwJobStat *job_stat = ts_bgw_job_stat_find(job->fd.id);
+	ScheduledBgwJob *sjob = MemoryContextAlloc(mctx, sizeof(ScheduledBgwJob));
+
+	*sjob = (ScheduledBgwJob)
+	{
+		.job = job,
+			.enabled = ts_bgw_job_stat_should_execute(job_stat, job),
+			.next_start = ts_bgw_job_stat_next_start(job_stat, job),
+	};
+
+	return sjob;
+}
+
+static bool
+scheduled_bgw_job_is_running(ScheduledBgwJob *sjob)
```

Comment:
I think this function name is misleading, ISTM this checks if the job was once started, regardless of whether it's running now.

Reply:
No it checks whether it's running now because scheduled_bgw_job_mark_stopped will set the handle to NULL.

Reply:
So compare this function to something like waitpid, that function will return information about the ground truth of the process state regardless of when and how it's called. This function returns information based on the last call to scheduled_bgw_job_mark_stopped, which, depending on where we are relative to scheduled_bgw_job_list_check_stopped_or_timed_out_jobs may or may not be up-to-date. A more accurate name would be something like scheduled_bgw_job_is_scheduled.